crates/script/src/progress.rs

```diff
@@ -0,0 +1,156 @@
+use crate::sequence::ScriptSequence;
+use alloy_chains::Chain;
+use alloy_primitives::B256;
+use foundry_cli::init_progress;
+use indicatif::{MultiProgress, ProgressBar, ProgressStyle};
+use parking_lot::RwLock;
+use std::{collections::HashMap, sync::Arc, time::Duration};
+use yansi::Paint;
+
+/// State of [ProgressBar]s displayed for the given [ScriptSequence].
+#[derive(Debug)]
+pub struct SequenceProgressState {
+    /// The top spinner with containt of the format "Sequence #{id} on {network} | {status}""
+    top_spinner: ProgressBar,
+    /// Progress bar with the count of transactions.
+    txs: ProgressBar,
+    /// Progress var with the count of confirmed transactions.
+    receipts: ProgressBar,
+    /// Standalone spinners for pending transactions.
+    tx_spinners: HashMap<B256, ProgressBar>,
+    /// Copy of the main [MultiProgress] instance.
+    multi: MultiProgress,
+}
+
+impl SequenceProgressState {
+    pub fn new(sequence_idx: usize, sequence: &ScriptSequence, multi: MultiProgress) -> Self {
+        let mut template = "{spinner:.green}".to_string();
+        template.push_str(
+            format!(" Sequence #{} on {}", sequence_idx + 1, Chain::from(sequence.chain)).as_str(),
+        );
+        template.push_str("{msg}");
```

Comment:
this doesn't format, use write! macro

Reply:
what do you mean? `.push_str("{msg}")` should not format, `{msg}` is a template key to use `set_message`/`with_message`

Reply:
oh right, you should still use write! instead of format! above